kernel/src/scan/log_replay.rs

```diff
@@ -37,80 +37,108 @@ struct LogReplayScanner {
+        Ok(!self.already_seen(path, dv_unique_id) && is_add)
+    }
+}
+impl<'seen> RowVisitorBase for AddRemoveDedupVisitor<'seen> {
     fn selected_column_names_and_types(&self) -> (&'static [ColumnName], &'static [DataType]) {
         // NOTE: THe visitor assumes adds always come first, with removes optionally afterward
-        static ALL_NAMES_AND_TYPES: LazyLock<ColumnNamesAndTypes> = LazyLock::new(|| {
-            let (add_names, add_fields) = AddVisitor::names_and_types();
-            let (remove_names, remove_fields) = RemoveVisitor::names_and_types();
-            let names = add_names.iter().chain(remove_names).cloned().collect();
-            let fields = add_fields.iter().chain(remove_fields).cloned().collect();
-            (names, fields).into()
+        static NAMES_AND_TYPES: LazyLock<ColumnNamesAndTypes> = LazyLock::new(|| {
+            const STRING: DataType = DataType::STRING;
+            const INTEGER: DataType = DataType::INTEGER;
+            let types_and_names = vec![
+                (STRING, column_name!("add.path")),
+                (STRING, column_name!("add.deletionVector.storageType")),
+                (STRING, column_name!("add.deletionVector.pathOrInlineDv")),
+                (INTEGER, column_name!("add.deletionVector.offset")),
+                (STRING, column_name!("remove.path")),
+                (STRING, column_name!("remove.deletionVector.storageType")),
+                (STRING, column_name!("remove.deletionVector.pathOrInlineDv")),
+                (INTEGER, column_name!("remove.deletionVector.offset")),
+            ];
+            let (types, names) = types_and_names.into_iter().unzip();
+            (names, types).into()
         });
+        let (names, types) = NAMES_AND_TYPES.as_ref();
         if self.is_log_batch {
-            ALL_NAMES_AND_TYPES.as_ref()
+            (names, types)
         } else {
             // All checkpoint actions are already reconciled and Remove actions in checkpoint files
-            // only serve as tombstones for vacuum jobs. So no need to load them here.
-            AddVisitor::names_and_types()
+            // only serve as tombstones for vacuum jobs. So we only need to examine the adds here.
+            (&names[..4], &types[..4])
         }
     }
 
     fn visit<'a>(&mut self, row_count: usize, getters: &[&'a dyn GetData<'a>]) -> DeltaResult<()> {
-        let expected_getters = if self.is_log_batch { 29 } else { 15 };
+        let expected_getters = if self.is_log_batch { 8 } else { 4 };
         require!(
             getters.len() == expected_getters,
             Error::InternalError(format!(
-                "Wrong number of AddRemoveVisitor getters: {}",
+                "Wrong number of AddRemoveDedupVisitor getters: {}",
                 getters.len()
             ))
         );
+
+        // TODO(zach): Add a check that the selection vector never tries to skip a remove?
```

Comment:
I'm not sure it matters? Currently, we do not attempt data skipping on Remove actions, so the initial selection vector should never delete a Remove. Even if we started skipping on removes, the worst that can happen is we fail to skip some add that lacked stats. But how would the remove have stats if the add lacked them? Somebody would have had to do:
```
v0 - add (without stats)
v1 - add (with stats)
v2 - remove (with stats)
```
In that case, data skipping could eliminate the v1 and v2 actions while leaving v0. Which would cause us to unnecessarily read that file. A waste of work, but still correct because the WHERE clause that allowed us to skip v1 and v2 actions would also eliminate every row from the file.

Reply:
thanks it sounds like we can remove this then? agree probably overly cautious!